agent/lib/kontena/actors/container_exec.rb

```diff
@@ -33,22 +33,45 @@ def input(input)
       @write_pipe = nil
     end
 
+    # @param [Hash] size
+    # @option size [Integer] width
+    # @option size [Integer] height
+    def tty_resize(size)
+      return unless @container_exec
+
+      begin
+        @container_exec.resize({
+          'w' => size['width'], 'h' => size['height']
+        })
+        info "tty resized to #{size['width']}x#{size['height']}"
+      rescue Docker::Error::NotFoundError => exc
+        if exc.message.include?('process not found')
+          sleep 0.1
+          retry if @container_exec
```

Comment:
> Without weird retry logic docker will say this: `Docker::Error::NotFoundError: containerd: process not found for container`

Does that happen if you `resize` before the exec has been `start!`ed? That only happens later in the `defer` block, so it's entirely possible for `tty_resize` run before the call to `start!`... although it's also hard to tell when the exec has actually started :confused: 

What happens if there's a `resize` after the exec has stopped? I'm worried about this getting stuck in an infinite retry loop.

Reply:
It sets `@container_exec` to `nil` in `ensure`. Should be safe?

Reply:
Yeah, this looks very much like a race between the call to `@container_exec.resize` and the call to `@container_exec.start!`. In this case, it seems likely for the `resize` to happen before the `start!`:

```
D, [2017-09-19T09:31:40.700960 #1] DEBUG -- Kontena::RpcServer: rpc request: Kontena::Rpc::DockerContainerApi#create_exec ["b2cdc7e4da79574882f89d02fc582f0e23fead6e3c3b3e35fbac1fda7a09df5e"]
I, [2017-09-19T09:31:40.706165 #1]  INFO -- Kontena::Actors::ContainerExec: initialized (session 8b781dac-82e0-4298-a999-1741d5db6964)
D, [2017-09-19T09:31:40.727809 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#run_exec ["8b781dac-82e0-4298-a999-1741d5db6964", ["bash"], true, true]
I, [2017-09-19T09:31:40.728268 #1]  INFO -- Kontena::Actors::ContainerExec: starting command: ["bash"] (tty: true, stdin: true)
D, [2017-09-19T09:31:40.739328 #1] DEBUG -- Kontena::Workers::ServicePodWorker: container event: exec_create: bash 
D, [2017-09-19T09:31:40.742052 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_resize ["8b781dac-82e0-4298-a999-1741d5db6964", {"width"=>140, "height"=>39}]
I, [2017-09-19T09:31:40.748308 #1]  INFO -- Kontena::Actors::ContainerExec: resize tty to 140x39
D, [2017-09-19T09:31:40.749280 #1] DEBUG -- Kontena::Actors::ContainerExec: start exec 82b54c9c4779265a0ac18467aea7f25188440d2f17e8831ce7c8ab47f6c92406 with tty...
D, [2017-09-19T09:31:40.793657 #1] DEBUG -- Kontena::Workers::ServicePodWorker: container event: exec_start: bash 
E, [2017-09-19T09:31:40.871032 #1] ERROR -- Kontena::Actors::ContainerExec: containerd: process not found for container
D, [2017-09-19T09:31:40.960604 #1] DEBUG -- Kontena::Actors::ContainerExec: tty exec output: "root@redis-1:/data# "
I, [2017-09-19T09:31:41.030233 #1]  INFO -- Kontena::Actors::ContainerExec: resize tty to 140x39
```

However, this race is tricky to fix, because it's not enough to wait for the call to `start!` to happen, the actual `POST /exec/.../start` request has to go through... Not possible to know when that has happened with the `docker-api` client. It looks like the Docker CLI waits for the `POST /exec/.../start` [request](https://github.com/docker/cli/blob/139fcd3ee95f37f3ac17b1200fb0a63908cb6781/cli/command/container/exec.go#L127) to complete in some form before doing the `POST /exec/.../resize`, and that might even be a `Connection: upgrade` `Upgrade: tcp` request with a [HTTP 101 Switching Protocols](https://github.com/moby/moby/blob/5c57ca17d50c392b7ad627fc1215d7a7e91e0c41/client/hijack.go#L183) response...

The `retry` with the `ensure` -> `@container_exec = nil` is an acceptable workaround, but IMO this kind of error handling retry code really needs some kind of comment explaining what happens and why this is needed... more than just the GitHub PR discussion.

Reply:
TLDR: not feasible to fix the race condition, this `retry` is a good workaround, but it needs a comment explaining what's going on